Add sdt support

s-weigand opened the pull request:
* added custom base class for high dimentional data `glotaran.model.dataset.HighDimentionalDataset`
* added custom class for FLIM data `glotaran.model.specialized_datasets.FLIMDataset`
* implemented `pd.DataFrame` as intermediate repressentation for Datasets, IMO this will make it easier for to implement readers for custom formats in the future
* in `glotaran.dataio.file_readers`
  - added `df_to_SpectralTemporalDataset`
  - added `df_to_FLIMDataset`
  - added `sdt_to_df`
  - added `read_sdt`
* added support for FLIM legacy data (glotaran 1.5), `glotaran.dataio.legacy_readers.FLIM_legacy_to_df`
* added mapper for FLIM data to pixel position (tuple) `glotaran.dataio.mapper.get_pixel_map`
* pinned down versions of setup - and install requirements in setup.py

**Testing**

* added 194 tests for the readers (droping some of the `spectral_unit` and `time_unit` will greatly reduce the amount in tests and that for make the test suite pass faster, but that open for discussion)
* didn't add tests for the new Datasets since they will be totally rewritten by PR #150 anyway

**Closing issues**

closes #124 #142 #143 #144 #64

jsnel commented:
Please rename glotaran.model.dataset.HighDimentionalDataset to glotaran.model.dataset.HighDimensionalDataset

joernweissenborn commented on glotaran/dataio/file_readers.py:
```diff
@@ -0,0 +1,378 @@
+from ..model.dataset import DimensionalityError
+
+
+def df_to_SpectralTemporalDataset(input_df: pd.DataFrame, dataset_label: str,
```
In glotaran we prefer verbose names. please change to `dataframe_to_...`

joernweissenborn commented on glotaran/model/dataset.py:
```diff
@@ -49,96 +58,198 @@ def get_calculated_axis(self):
         raise NotImplementedError
 
-    def get_axis(self, label):
+    def get_axis(self, axis_label: str) -> Union[list, np.ndarray]:
```
the axis MUST be ndarrays

s-weigand replied:
another idea would be to make `get_axis` and `set_axis` privat ( `_get_axis` / `_set_axis` ) and let the user only see the custom implementation, as is done in `SpectralTemporalDataset` with `time_axis` and `spectral_axis`.

joernweissenborn replied:
I am generally not caring so much. I have set it now to ndarray, since I use excplicitly things like `axe.size`. The thing is, that an axe in our context must have a size which can be retrieved by a standard method somehow and it must contain hashable items (actually only for the calculated axis ('time' in terms of a kinetic model).

Leaving that aside, `Dataset` is an abstract class without ABC. It provides a standart implementation for convinience (escpecially simulations). I don't want to strictly split it into interface and implementation, but maybe I should. In that case, I would leave the setter for axies out and only put the getter into. I experimentated with abstract properties, that would be a nice way to go.

As for this PR. consider Dataset as an interface like that:

```python
class Dataset:
    get_axis(label: str) -> ndarray
    data() -> ndarray
```

joernweissenborn commented on glotaran/model/dataset.py:
```diff
@@ -1,42 +1,51 @@
+    pass
+
+
 class Dataset(object):
```
This class have undergone alot of changes. you will have to adapt.

s-weigand replied:
Guess by that you mean a rebase after PR #150 was merged?

joernweissenborn commented on glotaran/model/dataset.py:
```diff
@@ -49,96 +58,198 @@ def get_calculated_axis(self):
         return lsv, svals, rsv.T
 
 
+class HighDimensionalDataset(Dataset):
```
This is a helper class and does not concern models. Move to somwhere in `dataio`

joernweissenborn commented on glotaran/model/specialized_datasets.py:
```diff
@@ -0,0 +1,123 @@
+"""
```
This also belongs to dataio.

s-weigand commented:
* moved docs linkcheck to a different test and made it `allowed to fail` in travis, since it depends on eternal resources

joernweissenborn commented:
Agreed

Joris Snellenburg <notifications@github.com> schrieb am Mo., 10. Sep. 2018,
09:25:

> glotaran.dataio should only hold actual file reading/writing classes
> a new module glotaran.datasets can be added for dataset implementation
> such as HighDimentionalDataset
> glotaran.model should remain agnostic to any and all specifica dataset
> implementations.
> @s-weigand <https://github.com/s-weigand> please made changes accordingly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/glotaran/glotaran/pull/151#issuecomment-419814225>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACJvwaoCf-nYZpItEn-r2YhsxPR_H3Xuks5uZhPXgaJpZM4WWzQ0>
> .
>

s-weigand commented:
* rebased to the current version (just after PR #153 merge)
* moved all datasets to a package `glotaran.datasets` (might need some discussion about that, since now the datasets and their discriptor are separated)
* removed `tox` from `.travis.yml`
* the 'docs-fix' is just a hot fix, for the docs to generate them at all (the underlying problem is that some imports fail in `glotaran.analysis` and the new version of sphinx does't like the hack of @joernweissenborn to use the string version of the importpath )
